### ISSUE TITLE
`WebExtEventBase` should not have different types for TAddListener and TCallback.

### ISSUE THREAD

**Issue by MicahZoltu** [jsmnbom/definitelytyped-firefox-webext-browser#20]
```typescript
interface WebExtEventBase<TAddListener extends (...args: any[]) => any, TCallback> {
    addListener: TAddListener;

    removeListener(cb: TCallback): void;

    hasListener(cb: TCallback): boolean;
}
```

`addListener`, `removeListener` and `hasListener` all operate on the same parameter type.  `WebExtEventBase` for some reason has a different type definition for `TCallback` and `the parameter for `addListener`.  This is resolved in `WebExtEvent` type alias, but there is no need to have this weird two layer indirection thing.  `WebExtEventBase` is _incorrect_ as it stands, only made correct by nature of `WebExtEvent`.

Recommend fixing `WebExtEventBase` to take in a single generic type parameter, rename it to `WebExtEvent` and delete the `WebExtEvent` type alias.

**Comment by MicahZoltu** [comment]
It appears that `WebExtEventBase` is referenced outside of that type alias.  I contend that if you _ever_ pass two different types for `TAddListener` and `TCallback` then the program is incorrect, thus `WebExtEventBase` should not be written as it currently is as the type system should prevent programming errors when it can (and this is one of the times it can).

**Comment by jsmnbom** [comment]
Yea you're probably right :)

Is this something you'd be able to PR? If not I'll probably get around to it some time in the future :)

**Comment by MicahZoltu** [comment]
If it was just a matter of a pull request to DefinitelyTyped I probably could.  My understanding though is that the type definitions for this are auto-generated, so I would need to first understand how the generation process works and then update it to generate the correct type information.

**Comment by jsmnbom** [comment]
The reason that i ask is that I'm not entirely sure what you're suggesting webextevent (which is used for events that don't take any addiotional properties) and webexteventbase (which is used for event that need extra properties) should look like
Maybe if you could write out those type definitions I can reformat the generator to use them or smth? :)

**Comment by MicahZoltu** [comment]
Current `WebExtEventBase` (for reference):
```typescript
interface WebExtEventBase<TAddListener extends (...args: any[]) => any, TCallback> {
    addListener: TAddListener;
    removeListener(cb: TCallback): void;
    hasListener(cb: TCallback): boolean;
}
```

The problem here is that there is no situation where it is _correct_ to have `TAddListener` not be a function that takes a single `TCallback` parameter and returns `void`.  For example, the following implementation of that interface is correct according to the type system, but _incorrect_ according to reality.
```typescript
class Foo implements WebExtEventBase<() => void, () => void> {
    addListener(): void;
    removeListener(cb: () => any): void;
    hasListener(cb: () => any): boolean;
}
```
Similarly, the following is a more correct looking implementation that is still incorrect in reality:
```typescript
class Foo implements WebExtEventBase<(cb: () => void) => void, (a: string) => void> {
    addListener(cb: () => void): void;
    removeListener(cb: (a: string) => void): void;
    hasListener(cb: (a: string) => void): boolean;
}
```

I believe that `WebExtEventBase` should look like the following:
```typescript
interface WebExtEventBase<TCallback extends Function> {
    addListener(cb: TCallback): void;
    removeListener(cb: TCallback): void;
    hasListener(cb: TCallback): boolean;
}
```

This way, the user of the interface cannot create an invalid implementation of it.  If this change is made, then I believe that `WebExtEvent` type alias is no longer needed, as `WebExtEventBase` can just be used directly.

**Comment by jsmnbom** [comment]
Alright I see now yea. And I'd agree that that is indeed how a normal `WebExtEvent` should usually behave. The issue (and the reason that `WebExtEventBase` exists is because of events like https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/proxy/onRequest#Syntax and https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webNavigation/onBeforeNavigate#Syntax (despite part of that page actually being wrong) which need additional arguments to addListener.

**Comment by MicahZoltu** [comment]
Ah, I see.  My personal preference would be to just have 2-3 different interfaces, since I think they are different enough.  If you stick with having a single base interface, I think you could update it to the following which doesn't allow for mismatched add/remove methods.
```typescript
interface WebExtEventBase<TCallback extends Function> {
	addListener(cb: TCallback, ...args: any[]): void;
	removeListener(cb: TCallback): void;
	hasListener(cb: TCallback): boolean;
}
// example simple implementation
class Foo implements WebExtEventBase<(a: string) => void> {
	addListener(cb: (a: string) => void): void {}
	removeListener(cb: (a: string) => void): void {}
	hasListener(cb: (a: string) => void): boolean { return true }
}
// example complex implementation
class Bar implements WebExtEventBase<(a: string) => void> {
	addListener(cb: (a: string) => void, filter: (x: string) => boolean): void {}
	removeListener(cb: (a: string) => void): void {}
	hasListener(cb: (a: string) => void): boolean { return true }
}
```

**Comment by jsmnbom** [comment]
How does [this output](https://gist.github.com/jsmnbom/d87c315ac404de68cbc59c80fc06aa11) look, is that kinda what you mean?
(Interesting bits at [L2512-L2550](https://gist.github.com/jsmnbom/d87c315ac404de68cbc59c80fc06aa11#file-index-d-ts-L2512-L2550) and [L2577](https://gist.github.com/jsmnbom/d87c315ac404de68cbc59c80fc06aa11#file-index-d-ts-L2577))

**Comment by MicahZoltu** [comment]
Looks good!

Note: You can simplify the definition a bit by using `Function`:
```typescript
interface WebExtEvent<TCallback extends Function> {
    addListener(cb: TCallback): void;
    removeListener(cb: TCallback): void;
    hasListener(cb: TCallback): boolean;
}
```
The base TypeScript libraries contain a type definition for `Function` that essentially evaluates to the same thing as you had, but in a way that is a bit more clear to the reader (IMO).

**Comment by jsmnbom** [comment]
Ah yea, I had that originally, but the DefinitelyTyped linter doesn't like it, so I went back to the current version.

But I'll submit this version to DefinitelyTyped soon then and then you should be able to pull down the types without a couple of hours afterwards ^^

**Comment by MicahZoltu** [comment]
Will it include a fix for #19 as well?

**Comment by jsmnbom** [comment]
Yes, it will :)
